sprocs/instance_question_select_manual_grading_objects.sql

```diff
@@ -0,0 +1,45 @@
+-- BLOCK instance_question_select_last_variant_with_submission
+DROP FUNCTION IF EXISTS instance_question_select_manual_grading_objects(bigint);
+
+-- Retrieves the last variant for an instance question and last submission for the variant.
+
+CREATE OR REPLACE FUNCTION
+    instance_question_select_manual_grading_objects(
+        IN iq_id bigint,
+        OUT question jsonb,
+        OUT variant jsonb,
+        OUT submission jsonb
+    )
+AS $$
+BEGIN
+
+    SELECT to_jsonb(q.*)
+    INTO question
+    FROM
+        questions as q
+        JOIN assessment_questions AS aq ON (q.id = aq.question_id)
+        JOIN instance_questions AS iq ON (aq.id = iq.assessment_question_id)
+    WHERE iq.id = iq_id;
+
+    -- If a variant has not been found, student has not opened question
+    SELECT to_jsonb(v.*)
+    INTO variant
+    FROM
+        variants as v
+    WHERE v.instance_question_id = iq_id
+    ORDER BY v.date DESC, v.id DESC
+    LIMIT 1;
+
+    -- If variant is found but no submission, student did not submit anything
+    SELECT to_jsonb(s.*)
+    INTO submission
+    FROM
+        instance_questions AS iq
+        JOIN variants AS v ON (v.instance_question_id = iq.id)
+        JOIN submissions AS s ON (s.variant_id = v.id)
+    WHERE iq.id = iq_id
+    ORDER BY s.date DESC, s.id DESC
+    LIMIT 1;
```

Comment:
I believe this query could return a submission that doesn't match a variant for questions with multiple variants (e.g., if the last variant doesn't have submissions, but the first one has). You may want to review this to make a single select statement instead.

Reply:
Just want to unpack this a bit further @jonatanschroeder. I may be missing something here. Is there a scenario where we would want to manually grade the last variant without a submission if a variant prior to it has a submission? 

If there is no submission, the student did not submit an answer (ie. a student re-opens the instance question, where a new variant is produced, but a student decides not to answer again and sticks with the last submission as their final answer).

Reply:
I'm not saying that you should grade a non-submission, I'm saying that, in the situation where a new variant is created without a submission, this code would return a variant that doesn't match the submission. So for example, if the student receives variants 1, 2 and 3, submits an answer to variants 1 and 2, but not to 3, it would return the submission of 2, but the variant of 3.

Again, I'm not arguing about what needs to be graded, I'm saying that this code isn't returning elements that match each other. This is why I suggested a single select statement instead, which would both make it faster and returning matching values. For example, you could do (I haven't tested this, and the syntax may be a bit off, but it's just to illustrate):

```suggestion
    SELECT to_jsonb(q.*), to_jsonb(v.*), to_jsonb(s.*)
    INTO question, variant, submission
    FROM
        instance_questions AS iq
        JOIN assessment_questions AS aq ON (aq.id = iq.assessment_question_id)
        JOIN questions AS q ON (q.id = aq.question_id)
        LEFT JOIN variants AS v ON (v.instance_question_id = iq.id)
        LEFT JOIN submissions AS s ON (s.variant_id = v.id)
    WHERE iq.id = iq_id
    ORDER BY s.date DESC, s.id DESC
    LIMIT 1;
```

Reply:
Just explored this issue a little bit further. 

You probably already know this, but there is a difference between when variants are produced based on 'save' and 'save & grade' actions. The former action, which we are now concentrating on to support the manual grading feature, will not create a new variant after a submission. The latter will. 

I suspect after something is manually graded, a new variant will be created. Just confirming that now.

Reply:
I'm reading what you just posted now.

Reply:
Ah I see. I assumed that was invalid SQL syntax, but Postgresql certainly allows it. I'm much happier with that approach.